File: java/Heaps/SortAKSortedArray.java
import java.util.PriorityQueue;

public class SortAKSortedArray {
    public int[] sortAKSortedArray(int[] nums, int k) {
        // Populate a min-heap with the first k + 1 values in 'nums'.
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int i = 0; i < k + 1; i++) {
            minHeap.offer(nums[i]);
        }
        // Replace elements in the array with the minimum from the heap at each 
        // iteration.
        int insertIndex = 0;
        for (int i = k + 1; i < nums.length; i++) {
            nums[insertIndex] = minHeap.poll();
            insertIndex++;
            minHeap.offer(nums[i]);
        }
        // Pop the remaining elements from the heap to finish sorting the array.
        while (!minHeap.isEmpty()) {
            nums[insertIndex] = minHeap.poll();
            insertIndex++;
        }
        return nums;
    }
}

File: kotlin/Heaps/SortAKSortedArray.kt
import java.util.PriorityQueue

fun sortAKSortedArray(nums: IntArray, k: Int): IntArray {
    val nums = IntArray(nums.size)
    // Populate a min-heap with the first k + 1 values in 'nums'.
    val minHeap = PriorityQueue<Int>()
    for (i in 0 until k + 1) {
        minHeap.add(nums[i])
    }
    // Replace elements in the array with the minimum from the heap at each
    // iteration.
    var index = 0
    for (i in k + 1 until nums.size) {
        nums[index++] = minHeap.poll()
        minHeap.add(nums[i])
    }
    // Pop the remaining elements from the heap to finish sorting the array.
    while (minHeap.isNotEmpty()) {
        nums[index++] = minHeap.poll()
    }
    return nums
}
